File: validators/constraint_checker.py
"""
constraint_checker.py — static validation of a configuration against the
device database and the constraints table.

check(config, pinmux, constraints) -> list[findings]
Each finding: {severity, rule, message, pin?, function?}
ERROR findings block export; WARNING findings are advisory.
"""

from __future__ import annotations

VALID_AIO = {"AIO2", "AIO4", "AIO6", "AIO10", "AIO12", "AIO14"}
FIXED_TYPES = {"power", "ground", "analog_ref", "reset", "jtag", "reserved",
               "crystal", "power_ctrl"}
MAX_SYSCLK = 60_000_000  # 60 MHz
TBPRD_MAX = 0xFFFF
DB_MAX = 1023  # 10-bit DBRED/DBFED


def _f(sev, rule, msg, pin=None, function=None):
    d = {"severity": sev, "rule": rule, "message": msg}
    if pin is not None:
        d["pin"] = pin
    if function is not None:
        d["function"] = function
    return d


def check(config: dict, pinmux: dict, constraints: dict) -> list:
    findings: list = []
    pins_db = pinmux.get("pins", {})
    raw_pins = config.get("pins", {}) or {}
    if isinstance(raw_pins, dict):
        assigned = []
        for key, value in raw_pins.items():
            pin = dict(value or {})
            pin.setdefault("physical_pin", int(key))
            pin.setdefault("pin", pin["physical_pin"])
            assigned.append(pin)
    else:
        assigned = raw_pins
    params = config.get("params", {}) or {}
    wizard = config.get("wizard")

    # ── index assigned pins ──────────────────────────────────────────────
    seen_pin: dict[int, dict] = {}
    fn_to_pin: dict[str, int] = {}
    tz_assigned: dict[str, int] = {}

    for a in assigned:
        pnum = a.get("pin")
        fn = a.get("function")
        mux = a.get("mux")
        sig = a.get("signal")
        pdef = pins_db.get(str(pnum))

        # PIN_CONFLICT
        if pnum in seen_pin and seen_pin[pnum].get("function") != fn:
            findings.append(_f("ERROR", "PIN_CONFLICT",
                               f"引脚 {pnum} ({sig}) 同时被 "
                               f"'{seen_pin[pnum].get('function')}' 和 '{fn}' 占用",
                               pin=pnum, function=fn))
        seen_pin[pnum] = a

        if pdef is None:
            findings.append(_f("ERROR", "MUX_INVALID",
                               f"引脚 {pnum} 不在设备数据库中", pin=pnum))
            continue

        # POWER_PIN_GPIO / fixed pin
        if not pdef.get("configurable", False) or pdef.get("pin_type") in FIXED_TYPES:
            findings.append(_f("ERROR", "POWER_PIN_GPIO",
                               f"引脚 {pnum} ({pdef.get('primary_signal')}) 是固定功能脚"
                               f"（{pdef.get('pin_type')}），不能生成 GPIO/外设配置",
                               pin=pnum))
            continue

        # MUX_INVALID
        valid_mux = {m["mux"]: m for m in pdef.get("mux_options", [])}
        if mux not in valid_mux:
            findings.append(_f("ERROR", "MUX_INVALID",
                               f"引脚 {pnum} 不存在 MUX={mux}（可选：{sorted(valid_mux)}）",
                               pin=pnum))
        else:
            opt = valid_mux[mux]
            # function name mismatch
            if fn and opt["function"] != fn:
                findings.append(_f("WARNING", "MUX_INVALID",
                                   f"引脚 {pnum} MUX{mux} 的数据库功能是 "
                                   f"'{opt['function']}'，与配置 '{fn}' 不一致",
                                   pin=pnum, function=fn))
            # R3 separates the evidence questions.  source_verified remains a
            # read-only compatibility fallback for older databases.
            signal_verified = opt.get("signal_verified", opt.get("source_verified", False))
            mux_verified = opt.get("mux_value_verified", opt.get("source_verified", False))
            generator_supported = opt.get("generator_supported", mux == 0)
            if signal_verified is False:
                findings.append(_f("ERROR", "UNRESOLVED_PARAM",
                                   f"引脚 {pnum} MUX{mux}={opt['function']} 尚未由 "
                                   f"SPRS584Q Table 5-1 确认信号存在，禁止选择",
                                   pin=pnum, function=fn))
            elif mux_verified is False:
                findings.append(_f("ERROR", "MUX_VALUE_UNVERIFIED",
                                   f"引脚 {pnum} MUX{mux}={opt['function']} 的数值尚未核实，"
                                   "只阻断该项 MUX 代码生成",
                                   pin=pnum, function=fn))
            elif generator_supported is False:
                findings.append(_f(
                    "WARNING", "PINMUX_ONLY",
                    f"{opt['function']} 允许保存并生成 pinmux；当前版本不生成完整外设寄存器初始化",
                    pin=pnum, function=fn))

        if fn:
            fn_to_pin[fn] = pnum
            if fn.startswith("TZ") and fn.endswith("n"):
                base = fn  # e.g. TZ1n
                if base in tz_assigned and tz_assigned[base] != pnum:
                    findings.append(_f("ERROR", "PIN_CONFLICT",
                                       f"{base} 同时分配到引脚 {tz_assigned[base]} 和 {pnum}",
                                       pin=pnum, function=fn))
                tz_assigned[base] = pnum

    # ── ADC AIO conflict: pin used as digital GPIO but is an AIO analog pin ──
    for a in assigned:
        pnum = a.get("pin")
        pdef = pins_db.get(str(pnum), {})
        aio = pdef.get("aio")
        if aio and a.get("mux") == 0:
            findings.append(_f("ERROR", "ADC_AIO_CONFLICT",
                               f"引脚 {pnum} 是模拟/AIO 脚（{aio}），作为数字 GPIO 使用"
                               f"会与 ADC 冲突", pin=pnum))
        if aio and aio not in VALID_AIO:
            findings.append(_f("ERROR", "ADC_AIO_CONFLICT",
                               f"引脚 {pnum} 的 {aio} 不是有效数字 AIO"
                               f"（仅 AIO2/4/6/10/12/14）", pin=pnum))

    # ── wizard-specific checks ────────────────────────────────────────────
    if wizard == "epwm_complementary" or params.get("freq_hz") or params.get("dead_ns"):
        findings += _check_epwm(params, fn_to_pin)

    if wizard == "system_clock" or params.get("target_mhz"):
        findings += _check_clock(params)

    if wizard == "adc_soc" or params.get("soc") is not None:
        findings += _check_adc(params, assigned, pins_db)

    # ── R2: ePWM A/B pairing constraints ──────────────────────────────────
    findings += _check_pwm_pairs(config, pins_db)
    findings += _check_project_modules(config, assigned, pins_db)

    # unresolved marker anywhere
    if config.get("unresolved"):
        findings.append(_f("ERROR", "UNRESOLVED_PARAM",
                           "配置中仍存在 UNRESOLVED 参数，禁止导出"))

    if not findings:
        findings.append(_f("INFO", "OK", "未发现约束冲突"))
    return findings


def _epwm_of(fn):
    import re
    m = re.match(r"^(EPWM\d+)([AB])$", (fn or "").upper())
    return (m.group(1), m.group(2)) if m else (None, None)


def _check_pwm_pairs(config, pins_db):
    """R2 §6: complementary-pair validation driven by pwm_modules (explicit)."""
    out = []
    mods = config.get("pwm_modules", {}) or {}
    raw = config.get("pins", {}) or {}
    pin_list = list(raw.values()) if isinstance(raw, dict) else raw
    pins = {p.get("function"): p for p in pin_list}

    # 1) cross-module A/B assignment (EPWM1A + EPWM2B) -> ERROR
    by_group: dict[str, list] = {}
    for p in pin_list:
        mod, ch = _epwm_of(p.get("function"))
        if mod and p.get("pair_mode") == "complementary":
            by_group.setdefault(p.get("group") or mod, []).append((mod, ch, p))
    for gid, members in by_group.items():
        mods_in_group = {m[0] for m in members}
        if len(mods_in_group) > 1:
            out.append(_f("ERROR", "EPWM_PAIR_CROSS_MODULE",
                          f"互补组 {gid} 内出现不同模块：{sorted(mods_in_group)}"
                          f"（A/B 必须同属一个 EPWM 模块）", function=gid))
        # 2) complementary group missing A or B -> ERROR
        chs = {m[1] for m in members}
        if "A" not in chs or "B" not in chs:
            out.append(_f("ERROR", "EPWM_PAIR_INCOMPLETE",
                          f"互补组 {gid} 缺少 {'A' if 'A' not in chs else 'B'} 通道",
                          function=gid))
        # 3) dead-band zero in complementary -> ERROR
        for mod, ch, p in members:
            if p.get("dead_ns") is not None and float(p.get("dead_ns", 0)) <= 0:
                out.append(_f("ERROR", "PWM_DEADBAND_ZERO",
                              f"互补组 {gid} 死区为 0 — 直通短路风险", function=gid))
                break
        # 4) Trip action inconsistent across A/B -> ERROR
        trips = {m[2].get("trip") for m in members if m[2].get("trip") is not None}
        if len(trips) > 1:
            out.append(_f("ERROR", "EPWM_TRIP_MISMATCH",
                          f"互补组 {gid} 的 A/B Trip 动作不一致：{sorted(trips)}", function=gid))

    # 5) module declared complementary but only one channel present -> ERROR
    for name, m in mods.items():
        mode = m.get("mode", m.get("pair_mode"))
        if mode == "complementary":
            have = set((m.get("pins") or {}).keys())
            if m.get("pin_a") is not None:
                have.add("A")
            if m.get("pin_b") is not None:
                have.add("B")
            if "A" not in have or "B" not in have:
                out.append(_f("ERROR", "EPWM_PAIR_INCOMPLETE",
                              f"{name} 声明为互补输出但只配置了 "
                              f"{'、'.join(sorted(have)) or '无'} 通道", function=name))
    return out


def _check_project_modules(config, assigned, pins_db):
    """R3 module-level semantic checks; errors stay local to their module."""
    out = []
    clock = config.get("system_clock")
    sysclk = MAX_SYSCLK
    if clock:
        try:
            sysclk = int(clock.get("sysclk_hz", float(clock.get("target_mhz", 60)) * 1e6))
            if sysclk > MAX_SYSCLK:
                out.append(_f("ERROR", "PLL_ILLEGAL",
                              f"系统时钟 {sysclk}Hz 超过 F28034 上限 {MAX_SYSCLK}Hz"))
        except (TypeError, ValueError):
            out.append(_f("ERROR", "UNRESOLVED_PARAM", "system_clock 数值非法"))

    for name, module in (config.get("pwm_modules", {}) or {}).items():
        try:
            freq = float(module.get("frequency_hz"))
            mode = module.get("count_mode", "up_down")
            tbprd = sysclk / ((2.0 if mode == "up_down" else 1.0) * freq)
            if not 1 <= tbprd <= TBPRD_MAX:
                out.append(_f("ERROR", "PWM_TBPRD_OVERFLOW",
                              f"{name} 的 TBPRD={tbprd:.0f} 超出 1..{TBPRD_MAX}",
                              function=name))
        except (TypeError, ValueError, ZeroDivisionError):
            out.append(_f("ERROR", "UNRESOLVED_PARAM",
                          f"{name} frequency_hz 非法", function=name))
        try:
            duty = float(module.get("duty", 0.5))
            if not 0 < duty < 1:
                out.append(_f("ERROR", "PWM_DUTY_ILLEGAL",
                              f"{name} duty 必须在 0 和 1 之间", function=name))
        except (TypeError, ValueError):
            out.append(_f("ERROR", "PWM_DUTY_ILLEGAL",
                          f"{name} duty 非法", function=name))
        if module.get("mode") == "complementary":
            db = module.get("deadband") or {}
            if not db.get("enabled", True) or float(db.get("red_ns", 0)) <= 0 or \
                    float(db.get("fed_ns", 0)) <= 0:
                out.append(_f("ERROR", "PWM_DEADBAND_ZERO",
                              f"{name} 互补输出必须配置非零 RED/FED 死区", function=name))
        trip = module.get("trip") or {}
        if trip.get("enabled"):
            source = str(trip.get("source", "")).upper().rstrip("N")
            trip_pin = next((p for p in assigned
                             if str(p.get("function", "")).upper().rstrip("N") == source), None)
            if trip_pin is None:
                out.append(_f("ERROR", "PWM_TRIP_PIN_MISSING",
                              f"{name} 选择了 {source}，但 ProjectConfig 没有对应 TZ 物理脚",
                              function=name))

    adc = config.get("adc")
    if adc:
        try:
            if int(adc.get("acqps")) < 7:
                out.append(_f("WARNING", "ADC_ACQPS_TOO_SHORT",
                              f"ACQPS={adc.get('acqps')} 采样窗过短（<7 周期）"))
        except (TypeError, ValueError):
            out.append(_f("ERROR", "UNRESOLVED_PARAM", "ADC ACQPS 非法"))
    return out


def _check_epwm(params, fn_to_pin):
    out = []
    freq = params.get("freq_hz")
    dead = params.get("dead_ns")
    duty = params.get("duty")

    # TBPRD overflow (assume up-down counting, no extra divider)
    if freq:
        try:
            tbprd = MAX_SYSCLK / (2.0 * float(freq))
            if tbprd > TBPRD_MAX:
                out.append(_f("ERROR", "PWM_TBPRD_OVERFLOW",
                              f"频率 {freq}Hz 需要 TBPRD={tbprd:.0f}，超过 {TBPRD_MAX}。"
                              f"请提高分频或降低目标频率"))
        except (ValueError, ZeroDivisionError):
            out.append(_f("ERROR", "UNRESOLVED_PARAM", "freq_hz 非法"))

    # dead band
    if dead is not None:
        try:
            ticks = float(dead) / (1e9 / MAX_SYSCLK)  # ns / 16.67ns
            if float(dead) <= 0:
                out.append(_f("ERROR", "PWM_DEADBAND_ZERO",
                              "互补 PWM 死区为 0 — 上下管直通短路风险", function="DeadBand"))
            elif ticks > DB_MAX:
                out.append(_f("ERROR", "PWM_DBRED_OVERFLOW",
                              f"死区 {dead}ns ≈ {ticks:.0f} tick，超过 10bit 上限 {DB_MAX}"))
        except ValueError:
            out.append(_f("ERROR", "UNRESOLVED_PARAM", "dead_ns 非法"))

    if duty is not None:
        try:
            d = float(duty)
            if not (0.0 < d < 1.0):
                out.append(_f("WARNING", "UNRESOLVED_PARAM",
                              f"占空比 {d} 超出 (0,1) 区间"))
        except ValueError:
            pass

    # PWM without Trip Zone -> warning
    has_tz = any(fn.startswith("TZ") for fn in fn_to_pin)
    if not has_tz:
        out.append(_f("WARNING", "PWM_NO_TRIP",
                      "ePWM 未配置 Trip Zone — 无硬件过流/故障保护，功率级风险高"))
    return out


def _check_clock(params):
    out = []
    mhz = params.get("target_mhz")
    if mhz is not None:
        try:
            if float(mhz) > 60:
                out.append(_f("ERROR", "PLL_ILLEGAL",
                              f"目标 SYSCLK {mhz}MHz 超过 F28034 上限 60MHz"))
        except ValueError:
            out.append(_f("ERROR", "UNRESOLVED_PARAM", "target_mhz 非法"))
    return out


def _check_adc(params, assigned, pins_db):
    out = []
    soc = params.get("soc")
    acqps = params.get("acqps")
    # SOC duplicate
    socs = [a for a in assigned if a.get("soc") is not None]
    seen = {}
    for a in socs:
        s = a["soc"]
        if s in seen:
            out.append(_f("ERROR", "ADC_SOC_CONFLICT",
                          f"SOC{s} 被多个通道占用", function="ADC"))
        seen[s] = True
    if acqps is not None:
        try:
            if int(acqps) < 7:
                out.append(_f("WARNING", "ADC_ACQPS_TOO_SHORT",
                              f"ACQPS={acqps} 采样窗过短（<7 周期），采样可能不准"))
        except (ValueError, TypeError):
            pass
    return out

File: src/core/constraint_checker.js
/* Pure browser-side validation for the single R3.2 ProjectConfig schema. */
(function (root, factory) {
  const api = factory();
  if (typeof module === 'object' && module.exports) module.exports = api;
  root.ConstraintChecker = api;
})(typeof globalThis !== 'undefined' ? globalThis : this, function () {
  'use strict';

  const MAX_SYSCLK = 60000000;
  const TBPRD_MAX = 0xFFFF;
  const DB_MAX = 1023;

  function finding(severity, rule, message, extra = {}) {
    return { severity, rule, message, ...extra };
  }

  function pins(project) {
    return Object.entries(project?.pins || {}).map(([key, value]) => ({
      physical_pin: Number(value?.physical_pin ?? key),
      ...(value || {}),
    }));
  }

  function exactOption(pinmux, pin) {
    const def = pinmux?.pins?.[String(pin.physical_pin)];
    if (!def) return { def: null, option: null };
    const option = (def.mux_options || []).find(item =>
      Number(item.mux) === Number(pin.mux) &&
      String(item.function).toUpperCase() === String(pin.function).toUpperCase(),
    ) || null;
    return { def, option };
  }

  function validatePins(project, pinmux, findings) {
    const seen = new Map();
    for (const pin of pins(project)) {
      const pnum = pin.physical_pin;
      if (seen.has(pnum)) {
        findings.push(finding(
          'ERROR', 'PIN_CONFLICT',
          `Pin${pnum} 同时被 ${seen.get(pnum)} 和 ${pin.function} 占用`,
          { pin: pnum, function: pin.function },
        ));
      }
      seen.set(pnum, pin.function);
      const { def, option } = exactOption(pinmux, pin);
      if (!def) {
        findings.push(finding('ERROR', 'PIN_NOT_FOUND',
          `Pin${pnum} 不在器件数据库中`, { pin: pnum }));
        continue;
      }
      if (!def.configurable) {
        findings.push(finding('ERROR', 'FIXED_PIN',
          `Pin${pnum} ${def.primary_signal} 是固定功能脚`, { pin: pnum }));
        continue;
      }
      if (!option) {
        findings.push(finding(
          'ERROR', 'MUX_FUNCTION_MISMATCH',
          `Pin${pnum} 的 MUX${pin.mux}/${pin.function} 不是 golden 数据库中的同一选项`,
          { pin: pnum, function: pin.function },
        ));
        continue;
      }
      if (!option.signal_verified) {
        findings.push(finding(
          'ERROR', 'SIGNAL_UNVERIFIED',
          `Pin${pnum} ${option.function} 的信号归属没有 golden 证据`,
          { pin: pnum, function: option.function },
        ));
      }
      if (!option.mux_value_verified) {
        findings.push(finding(
          'ERROR', 'MUX_VALUE_UNVERIFIED',
          `Pin${pnum} ${option.function} 的 MUX 数值没有 golden 证据`,
          { pin: pnum, function: option.function },
        ));
      }
      if (!option.pin_config_supported) {
        findings.push(finding(
          'ERROR', 'PIN_CONFIG_UNSUPPORTED',
          `Pin${pnum} ${option.function} 尚不支持生成引脚配置`,
          { pin: pnum, function: option.function },
        ));
      } else if (!option.peripheral_init_supported) {
        findings.push(finding(
          'WARNING', 'PINMUX_ONLY',
          `${option.function} 只生成引脚复用；当前版本不生成该外设完整初始化`,
          { pin: pnum, function: option.function },
        ));
      }
    }
  }

  function systemClockHz(project, findings) {
    const clock = project?.system_clock;
    if (!clock || ['existing', 'keep_existing'].includes(clock.mode)) return MAX_SYSCLK;
    const hz = Number(clock.sysclk_hz ?? Number(clock.target_mhz) * 1000000);
    if (!Number.isFinite(hz) || hz <= 0 || hz > MAX_SYSCLK) {
      findings.push(finding(
        'ERROR', 'PLL_ILLEGAL',
        `系统时钟必须在 1..${MAX_SYSCLK}Hz，当前值为 ${clock.sysclk_hz ?? clock.target_mhz}`,
      ));
      return MAX_SYSCLK;
    }
    return hz;
  }

  function validatePwm(project, sysclk, findings) {
    const allPins = pins(project);
    for (const [name, module] of Object.entries(project?.pwm_modules || {})) {
      const label = name.toUpperCase();
      if (!/^EPWM[1-7]$/.test(label)) {
        findings.push(finding('ERROR', 'PWM_MODULE_UNSUPPORTED',
          `${name} 不是 F28034 支持的 ePWM 模块`, { function: name }));
        continue;
      }
      const freq = Number(module.frequency_hz);
      const countMode = module.count_mode || 'up_down';
      const tbprd = sysclk / ((countMode === 'up_down' ? 2 : 1) * freq);
      if (!Number.isFinite(tbprd) || tbprd < 1 || tbprd > TBPRD_MAX) {
        findings.push(finding(
          'ERROR', 'PWM_TBPRD_OVERFLOW',
          `${label} 的 TBPRD=${Number.isFinite(tbprd) ? Math.round(tbprd) : '非法'} 超出 1..${TBPRD_MAX}`,
          { function: label },
        ));
      }
      const duty = Number(module.duty);
      if (!(duty > 0 && duty < 1)) {
        findings.push(finding('ERROR', 'PWM_DUTY_ILLEGAL',
          `${label} 占空比必须在 0 和 1 之间`, { function: label }));
      }
      if (module.mode === 'complementary') {
        if (module.pin_a == null || module.pin_b == null) {
          findings.push(finding(
            'ERROR', 'EPWM_PAIR_INCOMPLETE',
            `${label} 互补模式必须同时配置 A/B`, { function: label },
          ));
        }
        const db = module.deadband || {};
        const red = Number(db.red_ns);
        const fed = Number(db.fed_ns);
        const redTicks = Math.round(red * sysclk / 1e9);
        const fedTicks = Math.round(fed * sysclk / 1e9);
        if (!db.enabled || !(red > 0) || !(fed > 0)) {
          findings.push(finding(
            'ERROR', 'PWM_DEADBAND_ZERO',
            `${label} 互补输出必须有非零 RED/FED 死区`, { function: label },
          ));
        } else if (redTicks > DB_MAX || fedTicks > DB_MAX) {
          findings.push(finding(
            'ERROR', 'PWM_DEADBAND_OVERFLOW',
            `${label} 死区计数超出 10-bit 上限 ${DB_MAX}`, { function: label },
          ));
        }
      }
      const trip = module.trip || {};
      if (trip.enabled) {
        const source = String(trip.source || '').toUpperCase().replace(/N$/, '');
        const match = allPins.find(pin =>
          String(pin.function || '').toUpperCase().replace(/N$/, '') === source &&
          pin.module === name && pin.role === 'trip');
        if (!match || Number(match.physical_pin) !== Number(trip.pin)) {
          findings.push(finding(
            'ERROR', 'PWM_TRIP_PIN_MISSING',
            `${label} 选择了 ${source}，但 ProjectConfig 没有对应的 Trip 物理脚`,
            { function: label },
          ));
        }
      } else {
        findings.push(finding(
          'WARNING', 'PWM_NO_HARDWARE_TRIP',
          `${label} 未启用硬件 Trip；禁止据此直接开启功率级`,
          { function: label },
        ));
      }
    }
  }

  function validateTimers(project, sysclk, findings) {
    for (const [name, timer] of Object.entries(project?.timers || {})) {
      if (name.toUpperCase() !== 'TIMER0') {
        findings.push(finding('ERROR', 'TIMER_UNSUPPORTED',
          `当前只支持 TIMER0，不能生成 ${name}`, { function: name }));
        continue;
      }
      const prd = Math.round(sysclk * Number(timer.period_us) / 1e6) - 1;
      if (!Number.isFinite(prd) || prd <= 0 || prd > 0xFFFFFFFF) {
        findings.push(finding('ERROR', 'TIMER_PERIOD_ILLEGAL',
          `${name} period_us 无法转换为有效 PRD`, { function: name }));
      }
    }
  }

  function validateAdc(project, findings) {
    if (!project?.adc) return;
    const soc = Number(project.adc.soc);
    const acqps = Number(project.adc.acqps);
    if (!Number.isInteger(soc) || soc < 0 || soc > 15) {
      findings.push(finding('ERROR', 'ADC_SOC_ILLEGAL', 'ADC SOC 必须在 0..15'));
    }
    if (!Number.isInteger(acqps) || acqps < 0 || acqps > 63) {
      findings.push(finding('ERROR', 'ADC_ACQPS_ILLEGAL', 'ADC ACQPS 必须在 0..63'));
    } else if (acqps < 7) {
      findings.push(finding(
        'WARNING', 'ADC_ACQPS_TOO_SHORT',
        `ACQPS=${acqps} 小于 7，采样窗可能过短`,
      ));
    }
  }

  function validateAnalogRoutes(project, pinmux, findings) {
    if (project?.adc) {
      let pin = Number(project.adc.physical_pin);
      let def = Number.isInteger(pin) ? pinmux?.pins?.[String(pin)] : null;
      if (!def) {
        def = Object.values(pinmux?.pins || {}).find(item =>
          (item.analog_paths || []).some(route =>
            route.type === 'adc_input' &&
            String(route.function).toUpperCase() ===
              String(project.adc.channel).toUpperCase()));
        pin = Number(def?.physical_pin);
      }
      const present = (def?.analog_paths || []).some(route =>
        route.type === 'adc_input' &&
        String(route.function).toUpperCase() ===
          String(project.adc.channel).toUpperCase());
      if (!present) {
        findings.push(finding(
          'ERROR', 'ADC_PHYSICAL_ROUTE_MISMATCH',
          `Pin${pin} 不提供 ${project.adc.channel} 官方模拟路径`,
        ));
      }
      if (project?.aio?.[String(pin)]) {
        findings.push(finding(
          'ERROR', 'ANALOG_AIO_CONFLICT',
          `Pin${pin} 不能同时选择 ADC 模拟模式和数字 AIO`,
        ));
      }
    }
    for (const [name, route] of Object.entries(project?.comparator_inputs || {})) {
      const pin = Number(route.physical_pin);
      const def = pinmux?.pins?.[String(pin)];
      const present = (def?.analog_paths || []).some(item =>
        item.type === 'comparator_input' &&
        String(item.function).toUpperCase() === String(name).toUpperCase());
      if (!present) {
        findings.push(finding(
          'ERROR', 'COMPARATOR_PHYSICAL_ROUTE_MISMATCH',
          `Pin${pin} 不提供 ${name} 官方比较器输入路径`,
        ));
      }
      if (project?.aio?.[String(pin)]) {
        findings.push(finding(
          'ERROR', 'ANALOG_AIO_CONFLICT',
          `Pin${pin} 不能同时选择 Comparator 模拟模式和数字 AIO`,
        ));
      }
    }
    for (const [key, aio] of Object.entries(project?.aio || {})) {
      const pin = Number(aio.physical_pin ?? key);
      const def = pinmux?.pins?.[String(pin)];
      if (!def?.aio_function ||
          String(def.aio_function.function).toUpperCase() !==
            String(aio.function).toUpperCase()) {
        findings.push(finding(
          'ERROR', 'AIO_PHYSICAL_ROUTE_MISMATCH',
          `Pin${pin} 不提供 ${aio.function} 官方 AIO 路径`,
        ));
      }
    }
  }

  function validateProject(project, pinmux, family) {
    const findings = [];
    if (!project || Number(project.schema_version) !== 1) {
      findings.push(finding(
        'ERROR', 'SCHEMA_VERSION',
        '只接受 schema_version=1 的 R3.2 ProjectConfig',
      ));
    }
    if (project?.device && pinmux?.device &&
        !String(pinmux.device).toUpperCase().startsWith(String(project.device).toUpperCase())) {
      findings.push(finding('ERROR', 'DEVICE_MISMATCH',
        `ProjectConfig 器件 ${project.device} 与数据包 ${pinmux.device} 不一致`));
    }
    validatePins(project || {}, pinmux || {}, findings);
    const sysclk = systemClockHz(project || {}, findings);
    if (Number(family?.max_sysclk_mhz || 60) * 1e6 < sysclk) {
      findings.push(finding('ERROR', 'PLL_ILLEGAL',
        `SYSCLK=${sysclk}Hz 超过器件系列上限`));
    }
    validatePwm(project || {}, sysclk, findings);
    validateTimers(project || {}, sysclk, findings);
    validateAdc(project || {}, findings);
    validateAnalogRoutes(project || {}, pinmux || {}, findings);
    if (project?.unresolved) {
      findings.push(finding('ERROR', 'UNRESOLVED_PARAM',
        'ProjectConfig 仍包含 unresolved 标记'));
    }
    if (!findings.length) {
      findings.push(finding('INFO', 'OK', '未发现约束冲突'));
    }
    const blocking = findings.filter(item => item.severity === 'ERROR');
    return { ok: blocking.length === 0, blocking, findings };
  }

  return { validateProject };
});
